### ISSUE TITLE
Add cross-fading support

### ISSUE THREAD

**Issue by emoon** [emoon/HippoPlayer#110]
This one is a bit tricky as some players (TFMX, UADE) doesn't support two player instances at the same time :( so it needs to be configured on when it can be done.

Setting should include when it should happen, which formats, disable, etc

**Comment by jessicajonesunitedkingdom** [comment]
But you must decide: do I want to make a player or a DJ-like program?
This feature is more going towards a DJ/mixing program, which is not the goal I think?

**Comment by jessicajonesunitedkingdom** [comment]
But it would be SUPER cool, since I don't think there are programs to mix all those formats!!!

**Comment by emoon** [comment]
No, I don't want to make a DJ problem :) In this case I would simply just start playing two songs at the same time and cross-fade between them. I would not try to do a "proper mix" with BPM matching or anything like that, it would be much harder.

**Comment by jessicajonesunitedkingdom** [comment]
I once been to nintenDOOM in which they played music using all kinds of 8 bit devices (like game boy) and mixed it
It was incredible!

https://www.youtube.com/watch?v=W_dmB_PdsAA

**Comment by emoon** [comment]
Ah yeah! In my case I will not do anything fancy because I think that is outside of the scope of this tool.

**Comment by sagamusix** [comment]
Since HippoPlayer doesn't really do real-time audio (apart from such things as changing volume / EQ / etc.), you could keep a buffer that's at least as long as the crossfade duration and pre-render as much data as the buffer can hold; then unload the current file, load the next file and start filling the buffer. Post-processing such as volume / EQ should probably still be applied the moment the buffer is sent to the audio device, to keep user changes snappy.

**Comment by emoon** [comment]
Yeah I have actually thought about having an option where it's possible to pre-render the next track (in the playlist) fully. This will allow to have pitch / time stretching on "all" formats such as great fast seeks. 

Of course this will take some memory so just decoding the mix part above as you suggest will be a good idea also. It also depends if the user has some of the visualizers enabled (such as tracker, scopes for each channel, etc) then the regular player routine would have to be used. It would of course be possible to cache this data as well but then in starts to get messy :)

**Comment by emoon** [comment]
I re-read your comment and yes, that is a great idea :) that would indeed handle the problem where some players aren't "safe"
